src/shared/varlink-idl.c

```diff
@@ -0,0 +1,1577 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "memstream-util.h"
+#include "strv.h"
+#include "terminal-util.h"
+#include "varlink-idl.h"
+#include "set.h"
+
+enum {
+        COLOR_SYMBOL_TYPE,   /* interface, method, type, error */
+        COLOR_FIELD_TYPE,    /* string, bool, … */
+        COLOR_IDENTIFIER,
+        COLOR_MARKS,         /* [], ->, ?, … */
+        COLOR_RESET,
+        _COLOR_MAX,
+};
+
+static int varlink_idl_format_all_fields(FILE *f, const VarlinkSymbol *symbol, int direction, const char *indent, const char *const colors[static _COLOR_MAX]);
+
+static int varlink_idl_format_enum_values(
+                FILE *f,
+                const VarlinkSymbol *symbol,
+                const char *indent,
+                const char *const colors[static _COLOR_MAX]) {
+
+        bool first = true;
+
+        assert(f);
+        assert(symbol);
+        assert(symbol->symbol_type == VARLINK_ENUM_TYPE);
+
+        for (const VarlinkField *field = symbol->fields; field->field_type != _VARLINK_FIELD_TYPE_END_MARKER; field++) {
```

Comment:
fields[] is an embedded array of structs. The array is terminated with a struct initialized to all zeroes (i.e. `{}`). We defined `_VARLINK_FIELD_TYPE_END_MARKER` to zero, so that it will be implicitly set as field type on this last zero initialized struct. we use something similar on other places too.